src/responsibleai/rai_analyse/_rai_insight_data.py

```diff
@@ -0,0 +1,475 @@
+        return self.raiinsight
+
+    def filter_from_cohort(self, cohort):
+        from erroranalysis._internal.cohort_filter import filter_from_cohort
+
+        analyzer = self.raiinsight.error_analysis._analyzer
+        return filter_from_cohort(analyzer,
+                                  filters=cohort,
+                                  composite_filters=[])
+
+    def get_filtered_dataset(self, cohort):
+        filtered_dataset = self.filter_from_cohort(cohort)
+        model = self.raiinsight.error_analysis._analyzer.model
+        features = self.raiinsight.error_analysis._analyzer.feature_names
+        return {
+            "filtered_dataset": filtered_dataset,
+            "y_pred": model.predict(filtered_dataset[features]),
+            "y_test": filtered_dataset["true_y"],
+            "filter_conditions": [c["column"] + c["method"] + str(c["arg"]) for c in cohort]
+        }
+
+    def get_y_pred(self):
+        return self.y_pred
+
+    def get_y_test(self):
+        return self.raiinsight.test[self.raiinsight.target_column].to_numpy()
+
+    def get_test(self):
+        return self.raiinsight.test
+
+    def get_fairlearn_grouped_metric(self, sensitive_feature, metric):
+        from fairlearn.metrics import MetricFrame
+        grouped_metric = MetricFrame(metrics=metric_func_map[metric],
+                                     y_true=self.get_y_test(),
+                                     y_pred=self.get_y_pred(),
+                                     sensitive_features=self.raiinsight.test[sensitive_feature].to_numpy())
+
+        return grouped_metric
+
+    def get_error_analysis_data(self, target_metric):
+        # recompute error analysis based on target metric if requested metric does not match existing metric
+        if self.raiinsight.error_analysis._analyzer._metric != target_metric:
+            self.raiinsight.error_analysis._analyzer._metric = target_metric
+            for c in self.raiinsight.error_analysis._ea_config_list:
+                c.is_computed = False
+            self.raiinsight.error_analysis._ea_report_list = []
+            self.raiinsight.error_analysis.compute()
+
+        return self.raiinsight.error_analysis.get_data()
+
+    def get_causal_data(self):
+        def visit(node, parents, leaves_collection):
+            import copy
+            if node.leaf:
+                leaf = {
+                    "n_samples": node.n_samples,
+                    "treatment": node.treatment,
+                    "parents": parents
+                }
+
+                leaves_collection.append(leaf)
+
+            else:
+                this_node = {
+                    "feature": node.feature,
+                    "right_comparison": node.right_comparison,
+                    "comparison_value": node.comparison_value
+                }
+
+                # dictionary assignment is by reference. Need deepcopy for by value assignment
+                node_left = copy.deepcopy(this_node)
```

Comment:
Is this going to be creating lots of copies of the dataset? I can see this exploding RAM usage when combined with the recursion.

Reply:
Each copy is only a (distilled down) single node of the original tree, not the full tree. So at max the working memory should be limited by the original tree, plus the allocations for each reference.